### PR TITLE
Refactorings

### PR DESCRIPTION
**Pull request by safesparrow**
* Update lots of comments
* Small local refactorings
* Update ghost dependency function to no longer add links to all files when a namespace is not found anywhere
* Change the name of the test flag `DumpGraph` -> `DumpCheckingGraph`

I didn't use Fantomas, because when I ran it, I got changes in 25 files and also in lines I didn't modify, so I wasn't sure which of those I was supposed to commit.

### CODE REVIEW

**Comment by nojaf** [comment]
I can't seem to build the project when checking out this branch:
image

Did you push everything?

**Comment by nojaf** [comment]
> I didn't use Fantomas, because when I ran it, I got changes in 25 files and also in lines I didn't modify, so I wasn't sure which of those I was supposed to commit.

If unrelated files changed, I might mean that you were invoking Fantomas from another location than the repository root. 
`TL;DR` there are known issues with the `.fantomasignore` file, it only works properly if you run `dotnet fantomas src -r` from the repository root.

**Comment by nojaf** [comment]
scenario `"A open statement that leads nowhere should link to every file that came above it."` failed locally for me. This test covers ghost dependencies specifically.

File: src/Compiler/Driver/GraphChecking/Types.fsi [hunk 1]
```diff
@@ -74,20 +86,25 @@ type internal FileContentQueryState =
...
+    /// Although the `A` path exists in the Trie, it does not contain any relevant definitions (beyond itself).
+    /// </example>  
     | NodeDoesNotExposeData
-    /// A node was found with one or more file links
+    /// A node was found with one or more files that contain relevant definitions required for type-checking.  
     | NodeExposesData of Set<FileIndex>
 
-type internal QueryTrie = LongIdentifier -> QueryTrieNodeResult
+/// A function for querying a Trie (the Trie is defined within the function's context)
+type internal TrieQueryFunc = LongIdentifier -> QueryTrieNodeResult
```
[the remark this answers is not in the document]
Comment (safesparrow) [reply src/Compiler/Driver/GraphChecking/Types.fsi:105]: Conversely `QueryTrie` to me means `a Trie of Queries` - type names are usually nouns.

**Comment by safesparrow** [comment]
>But what really disturbed me was that none of my local projects were running in graph mode. Because --deterministic+ is enabled by default and you need to explicitly turn that off before graph can kick in.
Only when setting --deterministic- will --test:GraphBasedChecking do anything.
This should be added for the methodOptions in CompliationTests.fs, but maybe we also want to print a warning when --deterministic+ is combined with --test:GraphBasedChecking.

This is something I should fix in the main branch (or here) - perhaps when the feature is enabled via the dedicated flag, it should stay enabled even in deterministic builds, and only the experimental flag should have the implicit disable possible.

File: src/Compiler/Driver/GraphChecking/DependencyResolution.fs [hunk 2]
```diff
@@ -126,104 +124,133 @@ let indicesUnderNode (node: TrieNode) : Set<FileIndex> =
...
+                Set.intersect result.FoundDependencies filesDefiningNamespace
 
-            if Set.isEmpty intersection then
-                // The partial open did not lead to anything
-                // In order for it to exist in the current file we need to link it
-                // to some file that introduces the namespace in the trie.
-                if Set.isEmpty children then
-                    // In this case not a single file is contributing to the opened namespace.
-                    // As a last resort we assume all files are dependent, in order to preserve valid code.
-                    [| 0 .. (fileIndex - 1) |]
```
[the remark this answers is not in the document]
Comment (nojaf) [reply src/Compiler/Driver/GraphChecking/DependencyResolution.fs:164]: Sorry, I missed that comment, take that failing test as an example:
```fsharp
        scenario
            "A open statement that leads nowhere should link to every file that came above it."
            [
                sourceFile
                    "A.fs"
                    """
namespace X
"""
                    Set.empty
                sourceFile
                    "B.fs"
                    """
namespace Y
"""
                    Set.empty
                sourceFile
                    "C.fs"
                    """
namespace Z

open X
"""
                    (set [| 0; 1 |])
            ]
```

The `Trie` will have a node for `X`, `Y` and `Z` but it won't expose any files.

image

Because `X` is a namespace that does not expose any types or has any nested modules, it won't contain the `index` of `A.fs`. Querying for path `"X"` will return `QueryTrieNodeResult.NodeDoesNotExposeData`. So we do know it exists but we don't attribute it to any file. In that scenario, we add all the files that came before it as a last resort.

[the remark this answers is not in the document]
Comment (nojaf) [reply src/Compiler/Driver/GraphChecking/DependencyResolution.fs:164]: Yes, I don't think this will occur much in the wild.
Extending `TrieNodeInfo.Namespace` would definitely work as an alternative.

File: src/Compiler/Driver/GraphChecking/Types.fsi (hunk 1 above)
[the remark this answers is not in the document]
Comment (nojaf) [reply src/Compiler/Driver/GraphChecking/Types.fsi:105]: `Query` could also be a verb, a la `"to query the trie"`, `QueryTrie` for short.
Anyway, this isn't the most relevant thing. If you think it is consistent with the rest of the F# compiler codebase then go for it.

**Comment by safesparrow** [comment]
I revert controversial changes and fixed the build.

>If unrelated files changed, I might mean that you were invoking Fantomas from another location than the repository root.
TL;DR there are known issues with the .fantomasignore file, it only works properly if you run dotnet fantomas src -r from the repository root.

I ran the following:
```
PS C:\projekty\fsharp\nojaf> dotnet tool run fantomas --version
Fantomas v5.0.3+f1ec0088672e4b075aab51d5b66a99c7ae538260

PS C:\projekty\fsharp\nojaf> git reset --hard
PS C:\projekty\fsharp\nojaf> dotnet tool run fantomas -r .
```
and this gives the following result:
```
PS C:\projekty\fsharp\nojaf> git status
On branch graph-tc-new-tweaks
Your branch is up to date with 'safesparrow/graph-tc-new-tweaks'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
        modified:   vsintegration/tests/FSharp.Editor.Tests/BraceMatchingServiceTests.fs
        modified:   vsintegration/tests/FSharp.Editor.Tests/BreakpointResolutionServiceTests.fs
        modified:   vsintegration/tests/FSharp.Editor.Tests/CompletionProviderTests.fs
        modified:   vsintegration/tests/FSharp.Editor.Tests/DocumentDiagnosticAnalyzerTests.fs
        modified:   vsintegration/tests/FSharp.Editor.Tests/DocumentHighlightsServiceTests.fs
        modified:   vsintegration/tests/FSharp.Editor.Tests/EditorFormattingServiceTests.fs
        modified:   vsintegration/tests/FSharp.Editor.Tests/FsxCompletionProviderTests.fs
        modified:   vsintegration/tests/FSharp.Editor.Tests/GoToDefinitionServiceTests.fs
        modified:   vsintegration/tests/FSharp.Editor.Tests/HelpContextServiceTests.fs
        modified:   vsintegration/tests/FSharp.Editor.Tests/Helpers/RoslynHelpers.fs
        modified:   vsintegration/tests/FSharp.Editor.Tests/Hints/HintTestFramework.fs
        modified:   vsintegration/tests/FSharp.Editor.Tests/Hints/InlineParameterNameHintTests.fs
        modified:   vsintegration/tests/FSharp.Editor.Tests/Hints/InlineTypeHintTests.fs
        modified:   vsintegration/tests/FSharp.Editor.Tests/IndentationServiceTests.fs
        modified:   vsintegration/tests/FSharp.Editor.Tests/LanguageDebugInfoServiceTests.fs
        modified:   vsintegration/tests/FSharp.Editor.Tests/QuickInfoProviderTests.fs
        modified:   vsintegration/tests/FSharp.Editor.Tests/QuickInfoTests.fs
        modified:   vsintegration/tests/FSharp.Editor.Tests/SemanticClassificationServiceTests.fs
        modified:   vsintegration/tests/FSharp.Editor.Tests/SignatureHelpProviderTests.fs
        modified:   vsintegration/tests/FSharp.Editor.Tests/SyntacticColorizationServiceTests.fs
```
I'll keep reverting those changes, but I can't see what I'm doing wrong.

**Comment by nojaf** [comment]
Oh, that is interesting that. I always run Fantomas on the `src` folder. `vsintegration` is listed in the `.fantomasignore` but I'm not picked up. I'm not sure if that is to be expected or not. Maybe they shuffled stuff around. I'll look into this.
